keps/sig-api-machinery/20190927-insecure-backend-proxy.md

```diff
@@ -0,0 +1,201 @@
+cannot verify the kubelet identity, *but* the kube-apiserver's client certificate is still valid so the kubelet can still
+verify the kube-apiserver.  In this condition, a cluster-admin may need to get pod logs to debug his cluster.
+
+### Goals
+
+1. Allow cluster-admins to get pod logs from kubelets with expired serving certificates.  This will include an API change
+and an addition argument to `kubectl log`
+
+### Non-Goals
+
+1. Allow any bidirectional traffic proxied to kubelets.  This may be a future objective, but is not in scope for the current KEP.
+
+## Proposal
+
+In [PodLogOptions](https://github.com/kubernetes/api/blob/d58b53da08f5430bb0f4e1154a73314e82b5b3aa/core/v1/types.go), 
+add a `InsecureSkipTLSVerifyBackend bool`
+```go
+// PodLogOptions is the query options for a Pod's logs REST call.
+type PodLogOptions struct {
+	// ... existing fields snipped
+	
+	// insecureSkipTLSVerifyBackend indicates that the apiserver should not confirm the validity of the 
+	// serving certificate of the backend it is connecting to.  This will make the HTTPS connection between the apiserver
+	// and the backend insecure. This means that an attacker could intercept the logs request and provide false data.
+	// +optional
+	InsecureSkipTLSVerifyBackend bool `json:"insecureSkipTLSVerifyBackend,omitempty" protobuf:"varint,9,opt,name=insecureSkipTLSVerifyBackend"`
+}
+```
+The streamer for logs already prevents redirects, so an evil kubelet intercepting this traffic cannot redirect the kube-apiserver
+to use it's high powered credentials for a nefarious purpose.
+The `LocationStreamer` can take an additional argument and the `LogLocation`'s `NodeTransport` can actually produce a purpose
+built transport for insecure connections.
+
+To make this easier to use, we can add an `--insecure-skip-tls-verify-backend` flag to `kubectl log` which plumbs the option.
+This part of the KEP is a nice to have, since the kube-apiserver owns the backing capability which has general utility.
+
+### User Stories [optional]
+
+#### Story 1
+
+#### Story 2
+
+### Implementation Details/Notes/Constraints [optional]
+
+This design is safe based on the following conditions:
+
+1. kube-apiservers only authenticate to kubelets using certificates.  See `--kubelet-client-certificate`.  Certificate based
+authentication does not send replayable credentials to backends.
+2. Clients must opt-in to the functionality and the documentation must include the impact in plain english.
+3. Evil kubelets cannot trick kube-apiservers into using their credentials for any other purpose.
+Because the URL is chosen a the kube-apiserver and in order to use the kube-apiserver creds, the target of any proxy
+must terminate the connection, the evil kubelet cannot rewrite that destination URL.  The destination is the URL for
+getting logs for one particular pod.
+
+### Risks and Mitigations
+
```

Comment:
This would allow someone with write permissions to Node and Pod API objects and read permissions to Pod logs to make the API server exfiltrate data at `https://<nodeName>:<nodePort>/containerLogs/<podNamespace>/<podName>/<containerName>`, where all the bracketed parameters were under their control.

It would mean a kube-apiserver administrator could not set their kube-apiserver to successfully connect to make GET requests only to nodes in the same cluster, which is unfortunate.

In this opt-in insecure mode, would it be possible to verify the signer and hostname of the cert on the kubelet, but not the expiration time?

Reply:
I guess a similar ability is already present with an APIService configured to point to a service with `insecureSkipTLSVerify:true`, with a similar restriction on the path of requests sent to it (must have a leading `/apis/<group>/<version>` path prefix)

Reply:
I'm willing to have a look at custom tls handling if you like, but as you noted, this isn't the only way for a cluster-admin to direct kube-apiserver creds to specific urls.